codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java

```diff
@@ -184,6 +184,35 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Except
         }
     }
 
+    /**
+     * Tests a scenario when channel is closed while the handshake is in progress. Validates that the handshake
+     * future is notified in such cases.
+     */
+    @Test
+    public void testHandshakeFutureIsNotifiedOnChannelClose() throws Exception {
+        EmbeddedChannel clientChannel = createClientChannel(null);
+        EmbeddedChannel serverChannel = createServerChannel(null);
+
+        // Start handshake from client to server but don't complete the handshake for the purpose of this test.
+        transferAllDataWithMerge(clientChannel, serverChannel);
+
+        final WebSocketClientProtocolHandler clientWsHandler =
+                clientChannel.pipeline().get(WebSocketClientProtocolHandler.class);
+        final WebSocketClientProtocolHandshakeHandler clientWsHandshakeHandler =
+                clientChannel.pipeline().get(WebSocketClientProtocolHandshakeHandler.class);
+
+        final ChannelHandlerContext ctx = clientChannel.pipeline().context(WebSocketClientProtocolHandler.class);
+
+        // Close the channel while the handshake is in progress. The channel could be closed before the handshake is
+        // complete due to a number of varied reasons. To reproduce the test scenario for this test case,
+        // we would manually close the channel.
+        clientWsHandler.close(ctx, ctx.newPromise());
+
+        // At this stage handshake is incomplete but the handshake future should be completed exceptionally since
+        // channel is closed.
+        assertTrue(clientWsHandshakeHandler.getHandshakeFuture().isDone());
```

Comment:
Also call `clientChannel.finishAndReleaseAll()` and `serverChannel.finishAndReleaseAll()` otherwise we have leaks

Reply:
updated